ESQL: Limit how many bytes `concat()` can process

bpintea opened the pull request:
This adds a limit on the bytes length that a concatenated string can get to.
The limit is set to 1MB (per entry).
When the limit is hit, an exception is returned to the caller (similar to an accounting circuit breaking) and execution halted.

Related: #100288.

bpintea committed:
Limit how many bytes concat() can process

This adds a limit on the bytes length that a concatenated string can get
to.
The limit is set to 10MB (per entry).
When the limit is hit, an exception is returned to the caller (similar
to an accounting circuit breaking).

elasticsearchmachine commented:
Pinging @elastic/es-ql (Team:QL)

elasticsearchmachine commented:
Hi @bpintea, I've created a changelog YAML for you.

Commit:
Update docs/changelog/100360.yaml

elasticsearchmachine commented:
Pinging @elastic/elasticsearch-esql (:Query Languages/ES|QL)

bpintea committed:
Merge branch 'main' into fix/limit_concat_len

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/Concat.java:
```diff
@@ -83,13 +89,29 @@ public ExpressionEvaluator.Factory toEvaluator(Function<Expression, ExpressionEv
+            length += v.length;
+        }
+        if (length > MAX_CONCAT_LENGTH) {
+            throw new EsqlClientException("concatenating more than [" + MAX_CONCAT_LENGTH + "] bytes is not supported") {
```
thought: in my previous PRs, I used `VerificationException` instead of an anonymous `EsqlClientException`, so that we don't send back an exception that reads `type: ""` in the REST response.

bpintea replied:
`EsqlClientException.status()` returns `INTERNAL_SERVER_ERROR`, which is a 500 error. I thought of keeping the 4xx range.

bpintea replied:
EsqlClientException.status() returns a `INTERNAL_SERVER_ERROR`, which is a 500 error. I wanted to keep the 4xx range.
`VerificationException` does that too, but it seems a dedicated exception.

alex-spies reviewed (approved):
LGTM!

nik9000 commented on x-pack/plugin/esql/qa/server/single-node/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/single_node/HeapAttackIT.java:
```diff
@@ -168,11 +168,24 @@ public void testSmallConcat() throws IOException {
     public void testHugeConcat() throws IOException {
         initSingleDocIndex();
         assertCircuitBreaks(() -> concat(10));
     }
```
Can we just replace the body of this with `testHugeConcatRejected`.

bpintea replied:
Replaced.

nik9000 reviewed (approved):
It's probably worth a unit test as well.

bpintea committed:
Merge branch 'main' into fix/limit_concat_len

bpintea committed:
Adress review comments

- grow cratch builder with the checked length of input arguments
- replace disabled test

bpintea committed:
Lower max concat len to 1MB

Lower max concat len to 1MB

bpintea committed:
Lower max concat len to 1MB

Lower max concat len to 1MB in test.

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/Concat.java:
```diff
@@ -83,13 +89,29 @@ public ExpressionEvaluator.Factory toEvaluator(Function<Expression, ExpressionEv
     @Evaluator
     static BytesRef process(@Fixed(includeInToString = false) BreakingBytesRefBuilder scratch, BytesRef[] values) {
+        ensureIsBelowMaxLength(values);
         scratch.clear();
```
Suggestion applied.

nik9000 commented:
> Integrating this exception into the unit tests is doable, but not cleanly: either having a dedicated test ignoring the test providers (but being run against all of them), or contriving a test with warnings and type errors to avoid running most of the inherited tests. I've left it out for now, but I can add it in a follow up if needed.

I'd bang up a dedicated unit test for it. I think it's probably worth it, but I get that it's annoying with all the providers in the way.

bpintea committed:
Merge branch 'main' into fix/limit_concat_len

bpintea committed:
Add unit test

Add unit testing for max len concat'ing

bpintea committed:
Check rarely for oversize handling

Switch to lower frequency of oversized handling.

bpintea committed:
Stop checking after oversize test

Stop checking after oversize test

bpintea commented:
@elasticsearchmachine run elasticsearch-ci/part-2

bpintea commented:
@elasticsearchmachine run elasticsearch-ci/part-2

bpintea commented:
@elasticsearchmachine run elasticsearch-ci/part-2

bpintea committed:
Merge branch 'main' into fix/limit_concat_len

bpintea commented:
@elasticsearchmachine run elasticsearch-ci/part-3

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.11](https://github.com/elastic/elasticsearch/pull/100507)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/100507">](https://github.com/elastic/elasticsearch/pull/100507) |
